Neos.Media/Classes/Domain/Model/AssetSource/Neos/NeosAssetSource.php

```diff
@@ -0,0 +1,184 @@
+<?php
+namespace Neos\Media\Domain\Model\AssetSource\Neos;
+
+/*
+ * This file is part of the Neos.Media package.
+ *
+ * (c) Contributors of the Neos Project - www.neos.io
+  *
+ * This package is Open Source Software. For the full copyright and license
+ * information, please view the LICENSE file which was distributed with this
+ * source code.
+ */
+
+use Neos\Flow\Core\Bootstrap;
+use Neos\Flow\Http\HttpRequestHandlerInterface;
+use Neos\Flow\Http\Uri;
+use Neos\Flow\Mvc\ActionRequest;
+use Neos\Media\Domain\Model\AssetInterface;
+use Neos\Media\Domain\Model\AssetSource\AssetProxyRepositoryInterface;
+use Neos\Media\Domain\Model\AssetSource\AssetSourceInterface;
+use Neos\Media\Domain\Service\AssetService;
+use Neos\Media\Domain\Service\ThumbnailService;
+use Neos\Flow\Annotations as Flow;
+use Neos\Media\Exception\AssetServiceException;
+use Neos\Media\Exception\ThumbnailServiceException;
+use Psr\Http\Message\UriInterface;
+
+/**
+ * Asset source for Neos native assets
+ */
+final class NeosAssetSource implements AssetSourceInterface
+{
+    /**
+     * @var string
+     */
+    private $assetSourceIdentifier;
+
+    /**
+     * @var NeosAssetProxyRepositoryInterface
+     */
+    private $assetProxyRepository;
+
+    /**
+     * @Flow\Inject
+     * @var ThumbnailService
+     */
+    protected $thumbnailService;
+
+    /**
+     * @Flow\Inject
+     * @var AssetService
+     */
+    protected $assetService;
+
+    /**
+     * @Flow\Inject
+     * @var Bootstrap
+     */
+    protected $bootstrap;
+
+    /**
+     * @Flow\InjectConfiguration(path="asyncThumbnails")
+     * @var bool
+     */
+    protected $asyncThumbnails;
+
+    /**
+     * @param string $assetSourceIdentifier
+     * @param array $assetSourceOptions
+     */
+    public function __construct(string $assetSourceIdentifier, array $assetSourceOptions)
+    {
+        if (preg_match('/^[a-z][a-z0-9-]{0,62}[a-z]$/', $assetSourceIdentifier) !== 1) {
+            throw new \InvalidArgumentException(sprintf('Invalid asset source identifier "%s". The identifier must match /^[a-z][a-z0-9-]{0,62}[a-z]$/', $assetSourceIdentifier), 1513329665386);
```

Comment:
Non timestamp error code again - I've got a solution to generate timestamps for your PHPStorm if you are interested ;)

Reply:
Thanks ;-) Actually I use Phpstorm Live Templates for that:
![image](https://user-images.githubusercontent.com/95582/38977264-25f7de4a-43b4-11e8-8197-a1d0802dc257.png)
As far as I can tell, the wrong timestamps are from old code I refactored.

Reply:
Okay, you're right, actually the Phpstorm `timestamp()` function produces millisecond-based timestamps. And I never spotted the difference! I now switched to a plugin instead which produces second-based timestamps. Thanks, eagle-eyed Daniel!

Reply:
![image](https://user-images.githubusercontent.com/642226/38997053-ce295c5a-43ec-11e8-8bd1-ac5c2a474e73.png)